### ISSUE TITLE
Usage with React and/or CSS in JS

### ISSUE THREAD

**Issue by giocodes** [jhildenbiddle/css-vars-ponyfill#9]
Very cool work here @jhildenbiddle thanks for this project.

I would like to open the conversation about using the ponyfill on a React app with CSS-in-JS, to see if people have any suggestions on the best approach to apply it globally.

So far I've tried it at the root of my App, but it's not picking up children updates. Then I tried to run the ponyfill after listening for changes with the MutationObserver but that seems to be very expensive, to the point of freezing the DOM.

**Comment by jhildenbiddle** [comment]
Thanks for the kind words, @giocodes. Much appreciated.

I'm not familiar enough with React + CSS-in-JS to fully understand the problem, but I'm happy to dive in and see if we can get things working for you.

Since `css-vars-ponyfill` parses `<style>` and `<link>` elements, the React component would need to render one of these in order for the ponyfill to work as expected. If the styles are applied inline (e.g. `<div style="color: --primarycolor;">`), then the ponyfill will not work since inline styles are not transformed. Transformation of inline styles can be added to the ponyfill fairly easily, but this feature wasn't a priority for the initial release.

Regarding performance, we should first determine if the issues you ran into are a result of the ponyfill or the MutationObserver. My guess is that the freezing issue you mention is the result of an infinite loop rather than a performance issue. 

* Make sure your MutationObserver is only triggering the ponyfill when a `<link>` or `<style>` node is modified, added, or removed. 
* Ignore changes to the `<style id="css-vars-ponyfill">` element, which is the element created/updated by the ponyfill. Failing to do this will definitely cause an infinite loop: ponyfill adds/updates `<style id="css-vars-ponyfill">`, MutationObserver detects `<style>` change and triggers ponyfill, ponyfill adds/updates, MutationObserver triggers ponyfill... etc.)
* If a scenario exists where multiple `<style>` or `<link>` changes can occur as part of a single app update, be sure to [debounce](https://css-tricks.com/the-difference-between-throttling-and-debouncing/) the ponyfill calls to avoid unnecessary processing. For example, if two `<style>` elements were added to the DOM, a MutationObserver would trigger two separate mutation events, which would then call the ponyfill twice. Debouncing the ponyfill calls would ensure that all DOM changes are "done", then the ponyfill would be called once.

Keep in mind that MutationObserver is not supported below IE11. This may not be an issue for you, but it's the reason why I did not include a MutationObserver for handling `<style>` and `<link>` update in the lib.

Since you're using React, you may want to consider adopting a pattern that allows child component to pass state to their parent component(s). This would address the "[app] not picking up children updates" issue, remove the need for the MutationObserver, and address the lack of MutationObserver support in older browsers.

I hope this is helpful. Let me know if any of this addresses your issues. If not, we'll continue the discussion and come up with a solid path forward.

Thanks!

**Comment by giocodes** [comment]
@jhildenbiddle thank you for the very detailed response. 

I've put together a quick POC: https://stackblitz.com/edit/emotion-css-variables-ponyfill

Our current setup doesn't use inline styles, but I can see how that would be a great addition to the ponyfill. As you can see on the POC, we're using CSS-in-JS with Emotion which creates and appends `<style>` tags to the DOM's Head tag. Also, we are only aiming for EI11 so MutationObserver should be good.

I followed your advice and skipped the mutations on the style#css-vars-ponyfill element but the delay still happens. 

Even further, it happens without the MutationObserver too: notice that cssVars() is called once after the app finishes mounting (once React renders all components to the DOM), there is a snap until the styles are applied. 

About the idea of letting child components notify the parent app of updates, I've commented out the MutationObserver pattern and only left the cssVars() calls; once at the app lvl, and another on updates of the Toggle component, but the snap still happens. 
![ei11 css variables](https://user-images.githubusercontent.com/19365690/40027921-dd29be70-57a1-11e8-913e-cd592cc0476a.gif)

**Comment by jhildenbiddle** [comment]
Apologies for the delayed response, @giocodes. There have been three updates to the ponyfill in the last two days. Check out the [changelog](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/CHANGELOG.md) for details.

The good news is that I was able to dive into your POC (thanks for that!), determine the cause of the slowdown, and optimize the ponyfill for situations like yours. I also have some suggestion on how to leverage existing ponyfill options to optimize further. For what it's worth, the issues aren't related to React at all. React is the mechanism you've chosen to add/update stylesheets and call the ponyfill, but the same slowdowns could affect non-React projects.

First, let's make sure we're working reviewing demos in the same environment (Chrome) so we see accurate performance numbers:

- Open *Developer Tools*
- On the *Newtork* tab, check **Disable cache**
- On the *Application* tab, check **Bypass for network**

## [Demo 1](https://stackblitz.com/edit/css-vars-ponyfill-9-1)

This is a basic ponyfill demo. The demo shows what nodes are being processed and the first 256 characters of CSS for each. I'm intentionally leaving React out of the equation so we can just focus on the ponyfill. The setup is similar to your POC in two key ways:

* The use of `import 'semantic-ui-css/semantic.min.css'` for including Semantic UI's CSS.
* The use of `css-vars-ponyfill` @ v1.5.0

Performance (Chrome, macOS, 2009 Mac Pro @ 2.93GHz):

* Initial: 0.50s
* Updates: 0.32s - 0.40s

Performance suffers here because using `import` for Semantic UI results in a 637k of CSS being added to the DOM via a `<style>` tag. That is a *lot* of CSS, and by default the ponyfill processes *all* `<link>` and `<style>` nodes. The bottom line is that the ponyfill is being asked to do a lot of work--way more than it needs to. Since Semantic UI is so large and doesn't contain any CSS custom properties, we'd be better off if we could just exclude this CSS from the ponyfill (try not including Semantic UI's CSS and see what I mean). So...

## [Demo 2](https://stackblitz.com/edit/css-vars-ponyfill-9-2)

The code for this demo is identical the first with one exception:

* The ponyfill has been updated to v1.7.0

This version includes an optimization that prevents `<link>` or `<style>` CSS that does not contain a custom property declaration or function from being processed in the parse->transform->stringify pipeline. If there are no CSS custom properties, there's no reason to process the CSS. To be clear, the CSS is still included in the final output, but it's just inserted as-is without having any work done it. In the case of Semantic UI, that's 637k of CSS that the ponyfill doesn't have to deal with.

Performance (Chrome, macOS, 2009 Mac Pro @ 2.93GHz):

* Initial: 0.18s
* Updates: 0.07s - 0.12s

## [Demo 3](https://stackblitz.com/edit/css-vars-ponyfill-9-3)

This demo yields the best performance.

* The ponyfill has been updated to 1.7.0
* Semantic UI is included as a `<link>` instead of an `import` (see *external resources* under *Dependencies*)
* Semantic UI is excluded from ponyfill processing via the `exclude` option

The optimizations made in 1.7.0 are a big help, but we can further optimize by specifying `<link>` and `<style>` nodes to exclude entirely using the ponyfill's `exclude` option. This option has been available since launch, and it was designed to address exactly the issue you ran into with Semantic UI (the [documentation](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsexclude) uses Bootstrap as an example). The ponyfill will completely ignore nodes matched by the `exclude` CSS selector, so you can exclude by href (e.g. `link[href*=[semantic-ui"]`), by location in the DOM (e.g. `#app style`) or by attributes you add manually (e.g. `[data-cssvarsponyfill="true"]`). The CSS from these `exclude` nodes will not appear in the ponyfill-generated CSS output.

Performance (Chrome, macOS, 2009 Mac Pro @ 2.93GHz):

* Initial: 0.07s
* Updates: 0.05s

## [Demo 4](https://stackblitz.com/edit/css-vars-ponyfill-9-4)

This is a slightly modified version of your initial POC. The only difference is that I've added performance information to the console output. Click the console toggle in the lower-right corner of the StackBlitz window to view.

Performance (Chrome, macOS, 2009 Mac Pro @ 2.93GHz):

* Initial: 0.46s
* Updates: 0.27-0.43s

## [Demo 5](https://stackblitz.com/edit/css-vars-ponyfill-9-5)

Whew! We made it. This demo is your initial POC with the optimizations from Demo 3 and performance information in the console.

Performance (Chrome, macOS, 2009 Mac Pro @ 2.93GHz):

* Initial: 0.05s
* Updates: 0.01s

---

TL;DR:

1. Update to the latest version of the ponyfill
2. Exclude any CSS that doesn't contain CSS custom properties using the [`options.exclude`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsexclude).

Hope this helps!

**Comment by giocodes** [comment]
@jhildenbiddle thanks so much for such a complete and in-depth and response. I wanted to make sure I made enough time to review everything you found and it looks very promising. Cheers! 🙌 😄

**Comment by giocodes** [comment]
Also, just tested it on a EI11 vm and can confirm the improvement, amazing. 
![ei11 css-vars-ponyfill](https://user-images.githubusercontent.com/19365690/40864366-2a0313e8-65b9-11e8-8e01-5251a6ab6593.gif)

I'm curious if you also use vms or some similar tool to test older browsers, since on all the demos you mention Chrome.

**Comment by jhildenbiddle** [comment]
Happy to help, and thanks for the feedback. It helped improve the ponyfill for everyone.

As for testing, I use Chrome during development simply because the dev tools are better. I set `options.legacyOnly` to `false` so the ponyfill functions as if Chrome is a legacy browser, then once everything appears to be working I switch over and manually test on legacy browsers using VMs or BrowserStack.  The ponyfill also has automated tests that run on SauceLabs against a list of legacy browsers (see the list [here](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/4f69ca7218dfd2c51ffba56083414d24f3ffa5d5/karma.conf.js#L96-L139)).

Apologies if the Chrome performance numbers were confusing. I was using Chrome as a baseline instead of IE to remove virtual machine performance impact from the results.
